keps/sig-node/4580-deprecate-kubelet-runonce/README.md

```diff
@@ -0,0 +1,343 @@
+
+## Motivation
+
+Runonce mode doesn't support many newer pod features (init containers), and the pod lifecycle for runonce mode is even less well defined than normal.
+
+### Goals
+
+* Mark the `RunOnce` field in `KubeletConfiguration` and the `runonce` flag of kubelet as deprecated.
+* Remove kubelet support for RunOnce mode.
+
+### Non-Goals
+
+Runonce mode will be removed immediately. The deprecation and removal process will be gradual and feature gate to increase awareness among potential users.
+
+## Proposal
+
+This proposal outlines plans to deprecate and remove runonce mode in kubelet.
+
+### Risks and Mitigations
+
+## Design Details
+
+### KubeletConfiguration Change: KubeletConfiguration
+
+Mark the `RunOnce` field as Deprecated.
+
+### kubelet flag Change
+
+make the `runonce` flag as Deprecated.
+
+### Implement warning logging for runonce mode usage
+
+Starting in 1.31, during kubelet startup, if running in runonce mode, the kubelet will log a warning message, for example:
+
+```
+klog.Warning("runonce has been deprecated")
+```
+
+### Introduction RemoveRunonce feature gate
+
+With the introduction of the `RemoveRunonce` feature gate, Kubernetes aims to guide users through the deprecated runonce mode. When this feature gate is enabled, kubelet will refuse to start when the runonce flag is set.
+
+### Test Plan
+
+[x] I/we understand the owners of the involved components may require updates to
+existing tests to make this code solid enough prior to committing the changes necessary
+to implement this enhancement.
+
+##### Prerequisite testing updates
+
+##### Unit tests
+
+- N/A
+
+##### Integration tests
+
+- N/A
+
+##### e2e tests
+
+- N/A
+
+### Graduation Criteria
+
+#### Alpha
+
+- Feature gate `RemoveRunonce` is introduced, is disabled by default. Enabling this feature gate will fail the kubelet on startup with runonce mode enable.
+
+#### Beta
+
+- `RemoveRunonce` feature gate is enabled by default.
+- Failed when starting kubelet in runonce mode.
+
+#### GA
+
+- Remove `RemoveRunonce`  feature gate.
+- Remove the `RunOnce` field in `KubeletConfiguration` and the `runonce` flag of kubelet
+- Remove kubelet runonce mode.
+
+### Upgrade / Downgrade Strategy
+
+### Version Skew Strategy
+
+- N/A
+
+## Production Readiness Review Questionnaire
+
+### Feature Enablement and Rollback
+
+- [x] Feature gate (also fill in values in `kep.yaml`)
+  - Feature gate name: RemoveRunonce
+  - Components depending on the feature gate: kubelet
+  
+  - Will enabling / disabling the feature require downtime of the control
+    plane? Yes. Flag must be set on kubelet start. To disable, kubelet must be restarted. Hence, there would be brief control component downtime on a given node.
+  - Will enabling / disabling the feature require downtime or reprovisioning
+    of a node? Yes. See above; disabling would require brief node downtime.
+
+###### Does enabling the feature change any default behavior?
+
+No.
```

Comment:
```suggestion
Yes, this feature will fail kubelet if runonce is enabled.
```

Reply:
Eh. The feature will change the default behavior of runonce but not sure if that is what this question is asking actually..

Reply:
We did not change the "default behavior" because the default value for runonce mode is `false`.